File: html-redirect/redirect.cc
#include <iostream>
#include <string>
#include <filesystem>

std::string redirect(std::string inputPath, const std::string& basePath) {
    namespace fs = std::filesystem;

    if (inputPath.starts_with("/")) {
        inputPath = inputPath.substr(1);
    }

    // Resolve the full path
    fs::path fullPath = fs::path(basePath) / fs::path(inputPath);

    // 1. Try the given path if it's a file
    if (fs::is_regular_file(fullPath)) {
        return "/" + fs::relative(fullPath, basePath).string();
    }

    // 2. Try appending a `/` to the path and checking for an index.html
    fs::path indexPath = fullPath / fs::path("index.html");
    if (fs::is_regular_file(indexPath)) {
        // Also redirect
        return "/" + fs::relative(indexPath, basePath).string();
    }

    // 3. Check if appending a .html file extension gives a valid file
    fs::path htmlPath = fullPath;
    htmlPath.replace_extension(".html");
    if (fs::is_regular_file(htmlPath)) {
        return "/" + fs::relative(htmlPath, basePath).string();
    }

    // If no valid path is found, return empty string
    return "";
}

int main() {
    std::string basePath = "/Users/bret/Developer/ssc/socket/html-redirect/test-cases/1";
    std::string testPath = "/";
    std::string resolvedPath = redirect(testPath, basePath);
    std::cout << resolvedPath << std::endl;
}

File: html-redirect/redirect.js
import { access, stat } from 'fs/promises'
import path from 'path'

export async function redirect (inputPath, basePath) {
  // Resolve the full path for easier handling
  const fullPath = path.join(basePath, inputPath)

  // 1. Try the given path if it's a file
  if (await isFile(fullPath)) {
    return stripBasePath(fullPath, basePath)
  }

  // 2. Try appending a `/` to the path and checking for an index.html
  const indexPath = path.join(fullPath, 'index.html')
  if (await exists(indexPath)) return stripBasePath(indexPath, basePath)

  // 3. Check if appending a .html file extension gives a valid file
  const htmlPath = `${fullPath}.html`
  if (await isFile(htmlPath)) return stripBasePath(htmlPath, basePath)

  // Special case for root path
  if (inputPath === '/') {
    const rootIndexPath = path.join(basePath, 'index.html')
    return await exists(rootIndexPath) ? '/index.html' : null
  }

  // If no valid path is found, return null
  return null
}

// Helper function to check if a path exists
async function exists (filePath) {
  try {
    await access(filePath)
    return true
  } catch {
    return false
  }
}

// Helper function to check if a path points to a file
async function isFile (filePath) {
  try {
    const fileStat = await stat(filePath)
    return fileStat.isFile()
  } catch {
    return false
  }
}

// Helper function to strip the basePath from the fullPath
function stripBasePath (fullPath, basePath) {
  return fullPath.replace(basePath, '')
}
